python-checks/src/main/java/org/sonar/python/checks/GenericFunctionTypeParameterCheck.java

```diff
@@ -0,0 +1,69 @@
+/*
+ * SonarQube Python Plugin
+ * Copyright (C) 2011-2023 SonarSource SA
+ * mailto:info AT sonarsource DOT com
+ *
+ * This program is free software; you can redistribute it and/or
+ * modify it under the terms of the GNU Lesser General Public
+ * License as published by the Free Software Foundation; either
+ * version 3 of the License, or (at your option) any later version.
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
+ * Lesser General Public License for more details.
+ *
+ * You should have received a copy of the GNU Lesser General Public License
+ * along with this program; if not, write to the Free Software Foundation,
+ * Inc., 51 Franklin Street, Fifth Floor, Boston, MA  02110-1301, USA.
+ */
+package org.sonar.python.checks;
+
+import java.util.Optional;
+import org.sonar.check.Rule;
+import org.sonar.plugins.python.api.PythonSubscriptionCheck;
+import org.sonar.plugins.python.api.PythonVersionUtils;
+import org.sonar.plugins.python.api.SubscriptionContext;
+import org.sonar.plugins.python.api.symbols.Symbol;
+import org.sonar.plugins.python.api.tree.CallExpression;
+import org.sonar.plugins.python.api.tree.Name;
+import org.sonar.plugins.python.api.tree.Tree;
+import org.sonar.plugins.python.api.tree.TypeAnnotation;
+import org.sonar.python.tree.TreeUtils;
+
+@Rule(key = "S6796")
+public class GenericFunctionTypeParameterCheck extends PythonSubscriptionCheck {
+
+  private static String MESSAGE = "Use a generic type parameter for this function instead of a TypeVar.";
+
+  @Override
+  public void initialize(Context context) {
+    context.registerSyntaxNodeConsumer(Tree.Kind.PARAMETER_TYPE_ANNOTATION, GenericFunctionTypeParameterCheck::checkUseOfTypeVarGeneric);
+    context.registerSyntaxNodeConsumer(Tree.Kind.RETURN_TYPE_ANNOTATION, GenericFunctionTypeParameterCheck::checkUseOfTypeVarGeneric);
```

Comment:
As I re-read the specification, I admit this wasn't super clear, but I feel it would be better to have a similar approach to the one for classes. 

That is, we should subscribe to function definitions and collect their parameter and return type annotations, then raise a single issue on the function if some of them are `TypeVar`.  That would avoid raising multiple issues on the same function (it's really one issue: using legacy syntax) (and would make a quick fix possible if we ever decide to pursue that).

Reply:
Yes, this was my initial implementation actually. It is a little bit more lengthy, but I think the concept of raising one issue is valid. I will make the changes.

Reply:
I made the implementation change in 10dce7e8e2c4444e7cf8493ede52fc9d98f8e050.